rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingPropertyTest.scala

```diff
@@ -1,27 +1,115 @@
 package coop.rchain.rholang.interpreter.accounting
+import java.nio.file.Files
+
+import cats.Id
+import cats.effect.Sync
+import coop.rchain.crypto.hash.Blake2b512Random
+import cats.implicits._
+import coop.rchain.models._
+import coop.rchain.rholang.interpreter.Runtime.{RhoContext, RhoISpace}
+import coop.rchain.rholang.interpreter.errors.OutOfPhlogistonsError
 import coop.rchain.rholang.{PrettyPrinted, ProcGen}
-import coop.rchain.rholang.interpreter.Interpreter
-import coop.rchain.rholang.syntax.rholang_mercury.Absyn.Proc
+import coop.rchain.rholang.interpreter._
+import coop.rchain.rholang.syntax.rholang_mercury.Absyn.{PPar, Proc}
 import coop.rchain.rholang.syntax.rholang_mercury.PrettyPrinter
+import coop.rchain.rspace.history.Branch
+import coop.rchain.rspace.{Context, RSpace}
+import monix.eval.Task
 import org.scalacheck.Arbitrary
 import org.scalacheck.Test.Parameters
 import org.scalatest.{FlatSpec, Matchers}
 import org.scalatest.prop.PropertyChecks
 
+import scala.concurrent.duration._
+import monix.execution.Scheduler.Implicits.global
+
 class CostAccountingPropertyTest extends FlatSpec with PropertyChecks with Matchers {
+  import CostAccountingPropertyTest._
+
   implicit val params: Parameters = Parameters.defaultVerbose.withMinSuccessfulTests(1000)
   implicit val procArbitrary = Arbitrary(
     ProcGen.topLevelGen(5).map(PrettyPrinted[Proc](_, PrettyPrinter.print))
   )
 
+  implicit val taskExecutionDuration = 5.seconds
+
   def cost(proc: Proc): Cost = Cost(Interpreter.buildPar(proc).apply)
 
   behavior of "Cost accounting in Reducer"
 
-  it should "cost of substitution is never negative" in {
+  it should "have positive cost of substitution" in {
     forAll { p: PrettyPrinted[Proc] =>
       cost(p.value).value should be >= 0L
     }
+  }
+
+  behavior of "Charging phlos in Reducer"
+
+  it should "charge the same if the evaluation order is reversed" in {
+    forAll { (p: PrettyPrinted[Proc], q: PrettyPrinted[Proc]) =>
+      haveEqualResults(costOfExecution(p.value, q.value), costOfExecution(q.value, p.value))
+    }
+  }
 
+  it should "sequential execution and parallel execution have the same cost" in {
+    forAll { (p: PrettyPrinted[Proc], q: PrettyPrinted[Proc]) =>
+      haveEqualResults(
+        costOfExecution(new PPar(p.value, q.value)),
+        costOfExecution(q.value, p.value)
+      )
+    }
   }
 }
+
+object CostAccountingPropertyTest {
+  def haveEqualResults[A](tasks: Task[A]*)(implicit duration: Duration): Boolean =
+    tasks.toList
+      .sequence[Task, A]
+      .map { _.sliding(2).forall { case List(r1, r2) => r1 == r2 } }
```

Comment:
a foldLeft (or right) would be clearer here for me

Reply:
true!

Reply:
actually, when I tried to implement it I realised that the foldLeft approach is more complex and less readable